datadog/fwprovider/data_source_datadog_application_key.go

```diff
@@ -85,23 +90,48 @@ func (d *applicationKeyDataSource) Read(ctx context.Context, req datasource.Read
 			return
 		}
 		applicationKeysData := applicationKeysResponse.GetData()
-		if len(applicationKeysData) > 1 {
+		if len(applicationKeysData) > 1 && !state.ExactMatch.ValueBool() {
 			resp.Diagnostics.AddError("your query returned more than one result, please try a more specific search criteria", "")
 			return
 		}
 		if len(applicationKeysData) == 0 {
 			resp.Diagnostics.AddError("your query returned no result, please try a less specific search criteria", "")
 			return
 		}
-		applicationKeyPartialData := applicationKeysData[0]
-		id := applicationKeyPartialData.GetId()
-		applicationKeyResponse, _, err := d.Api.GetCurrentUserApplicationKey(d.Auth, id)
-		if err != nil {
-			resp.Diagnostics.AddError("error getting application key", "")
-			return
+		if state.ExactMatch.ValueBool() {
+			exact_matches := 0
+			for _, appKeyPartialData := range applicationKeysData {
+				appKeyAttributes := appKeyPartialData.GetAttributes()
+				if state.Name.ValueString() == appKeyAttributes.GetName() {
+					exact_matches++
+					id := appKeyPartialData.GetId()
+					ddResp, _, err := d.Api.GetCurrentUserApplicationKey(d.Auth, id)
+					if err != nil {
+						resp.Diagnostics.Append(utils.FrameworkErrorDiag(err, "error getting application key"))
+						return
+					}
+					appKeyData := ddResp.GetData()
+					d.updateState(&state, &appKeyData)
```

Comment:
Is it necessary to update the state no matter what the exact_match count is ?

Reply:
good catch, will update things to only updateState once at the end